Error while checking build/O2/fullCI for 745f724adb1abbcb5ffa1d8c79e71967a4794317 at 2023-08-25 17:04:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/QualityControl-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/O2Physics-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[0 more errors; see full log]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/11786/745f724adb1abbcb5ffa1d8c79e71967a4794317/build_O2_fullCI/pretty.html).

--- a/DataFormats/Detectors/EMCAL/include/DataFormatsEMCAL/Cell.h
+++ b/DataFormats/Detectors/EMCAL/include/DataFormatsEMCAL/Cell.h
@@ -14,6 +14,7 @@
 
 #include <bitset>
 #include <cfloat>
+#include <climits>
 #include "DataFormatsEMCAL/Constants.h"
 
 namespace o2
@@ -191,6 +192,12 @@ class Cell
   /// \brief Get encoded bit representation of cell type (for CTF)
   /// \return Encoded bit representation
   uint16_t getCellTypeEncoded() const;
+
+  static uint16_t encodeEnergyV0(float energy);
+  static uint16_t encodeEnergyV1(float energy, ChannelType_t celltype);
+  static uint16_t V0toV1(uint16_t energybits, ChannelType_t celltype);
+  static float decodeEnergyV0(uint16_t energybits);
+  static float decodeEnergyV1(uint16_t energybits, ChannelType_t celltype);
 
  private:
   /// \brief Set cell energy from encoded bit representation (from CTF)

--- a/DataFormats/Detectors/EMCAL/include/DataFormatsEMCAL/Cell.h
+++ b/DataFormats/Detectors/EMCAL/include/DataFormatsEMCAL/Cell.h
@@ -13,6 +13,7 @@
 #define ALICEO2_EMCAL_CELL_H_
 
 #include <bitset>
+#include <cfloat>
 #include "DataFormatsEMCAL/Constants.h"
 
 namespace o2
@@ -27,104 +28,92 @@ namespace emcal
 /// \since March 6, 2019
 /// \ingroup EMCALDataFormat
 ///
-/// # Base format for EMCAL cell information in the Compressed Timeframe
+/// # Cell content
 ///
 /// The cell class contains the relevant information for each tower per event
 /// - Tower ID
 /// - Energy of the raw fit
 /// - Time of the raw fit
 /// - Type of the cell
+///
+/// # Compression for CTF
+///
 /// While cell type and tower ID have a predefined range based on the hardware
 /// design, energy and time have a finite resolution influenced by the resolution
 /// of the digitizer. This is used in order to compress the information stored
 /// in the compressed timeframe by not storing the full double values but instead
 /// assigning a certain amount of bits to each information. Therefore for certain
-/// information (energy, time) precision loss has to be taken into account.
+/// information (energy, time) precision loss has to be taken into account. The number
+/// of bits assigned to each data member in the encoding are as follows:
 ///
-/// # Internal structure and resolution
-///
-/// The internal structure is a bit field compressing the information to
-/// 48 bits. The definition of the bit field as well as the value range and the resolution
-/// is listed in the table below:
-///
-/// | Bits  | Content       | Resolution    | Range                       |
-/// |-------|---------------|---------------|-----------------------------|
-/// | 0-14  | Tower ID      | -             | 0 to 17644                  |
-/// | 15-26 | Time (ns)     | 0.73 ns       | -600 to 900 ns              |
-/// | 27-40 | Energy (GeV)  | 0.0153 GeV    | 0 to 250 GeV                |
-/// | 41-42 | Cell type     | -             | 0=LG, 1=HG, 2=LEMon, 4=TRU  |
+/// | Content       | Number of bits |Resolution    | Range                       |
+/// |---------------|----------------|--------------|-----------------------------|
+/// | Tower ID      | 15             | -            | 0 to 17644                  |
+/// | Time (ns)     | 11             | 0.73 ns      | -600 to 900 ns              |
+/// | Energy (GeV)  | 14             | 0.0153 GeV   | 0 to 250 GeV                |
+/// | Cell type     | 2              | -            | 0=LG, 1=HG, 2=LEMon, 4=TRU  |
 ///
 /// The remaining bits are 0
 class Cell
 {
  public:
-  Cell();
-  Cell(short tower, float energy, float time, ChannelType_t ctype = ChannelType_t::LOW_GAIN);
+  /// \brief Default constructor
+  Cell() = default;
+
+  /// \brief Constructor
+  /// \param tower Tower ID
+  /// \param energy Energy
+  /// \param timestamp Cell time
+  /// \param ctype Channel type
+  Cell(short tower, float energy, float timestamp, ChannelType_t ctype = ChannelType_t::LOW_GAIN);
+
+  /// \brief Destructor
   ~Cell() = default; // override
 
-  void setTower(short tower) { getDataRepresentation()->mTowerID = tower; }
-  short getTower() const { return getDataRepresentation()->mTowerID; }
+  /// \brief Set the tower ID
+  /// \param tower Tower ID
+  void setTower(short tower) { mTowerID = tower; }
+
+  /// \brief Get the tower ID
+  /// \return Tower ID
+  short getTower() const { return mTowerID; }
 
   /// \brief Set the time stamp
-  /// \param time Time in ns
-  ///
-  /// The time stamp is expressed in ns and has
-  /// a resolution of 1 ns. The time range which can
-  /// be stored is from -1023 to 1023 ns. In case the
-  /// range is exceeded the time is set to the limit
-  /// of the range.
-  void setTimeStamp(float time);
+  /// \param timestamp Time in ns
+  void setTimeStamp(float timestamp) { mTimestamp = timestamp; }
 
   /// \brief Get the time stamp
   /// \return Time in ns
-  ///
-  /// Time has a resolution of 1 ns and can cover
-  /// a range from -1023 to 1023 ns
-  float getTimeStamp() const;
+  float getTimeStamp() const { return mTimestamp; }
 
   /// \brief Set the energy of the cell
   /// \brief Energy of the cell in GeV
-  ///
-  /// The energy range covered by the cell
-  /// is 0 - 250 GeV, with a resolution of
-  /// 0.0153 GeV. In case an energy exceeding
-  /// the limits is provided the energy is
-  /// set to the limits (0 in case of negative
-  /// energy, 250. in case of energies > 250 GeV)
-  void setEnergy(float energy);
+  void setEnergy(float energy) { mEnergy = energy; }
 
   /// \brief Get the energy of the cell
   /// \return Energy of the cell
-  ///
-  /// The energy is truncated to a range
-  /// covering 0 to 250 GeV with a resolution
-  /// of 0.0153 GeV
-  float getEnergy() const;
+  float getEnergy() const { return mEnergy; }
 
   /// \brief Set the amplitude of the cell
   /// \param amplitude Cell amplitude
-  ///
-  /// See setEnergy for more information
   void setAmplitude(float amplitude) { setEnergy(amplitude); }
 
   /// \brief Get cell amplitude
-  /// \return cell Amplitude
-  ///
-  /// Set getEnergy for more information
+  /// \return Cell amplitude in GeV
   float getAmplitude() const { return getEnergy(); }
 
   /// \brief Set the type of the cell
   /// \param ctype Type of the cell (HIGH_GAIN, LOW_GAIN, LEDMON, TRU)
-  void setType(ChannelType_t ctype) { getDataRepresentation()->mCellStatus = static_cast<uint16_t>(ctype); }
+  void setType(ChannelType_t ctype) { mChannelType = ctype; }
 
   /// \brief Get the type of the cell
   /// \return Type of the cell (HIGH_GAIN, LOW_GAIN, LEDMON, TRU)
-  ChannelType_t getType() const { return static_cast<ChannelType_t>(getDataRepresentation()->mCellStatus); }
+  ChannelType_t getType() const { return mChannelType; }
 
   /// \brief Check whether the cell is of a given type
   /// \param ctype Type of the cell (HIGH_GAIN, LOW_GAIN, LEDMON, TRU)
   /// \return True if the type of the cell matches the requested type, false otherwise
-  bool isChannelType(ChannelType_t ctype) const { return getType() == ctype; }
+  bool isChannelType(ChannelType_t ctype) const { return mChannelType == ctype; }
 
   /// \brief Mark cell as low gain cell
   void setLowGain() { setType(ChannelType_t::LOW_GAIN); }
@@ -154,36 +143,76 @@ class Cell
   /// \return True if the cell type is TRU, false otherwise
   Bool_t getTRU() const { return isChannelType(ChannelType_t::TRU); }
 
+  /// \brief Apply compression as done during writing to / reading from CTF
+  void truncate();
+
   void PrintStream(std::ostream& stream) const;
 
-  /// used for CTF encoding/decoding: access to packed data
-  void setPacked(uint16_t tower, uint16_t t, uint16_t en, uint16_t status)
+  /// \brief Initialize cell class from bit representation (for CTF decoding)
+  /// \param towerIDBits Encoded tower ID
+  /// \param timestampBits Encoded timestamp
+  /// \param energyBits Encoded energy
+  /// \param celltypeBits Encoded cell type
+  void initialiseFromEncoded(uint16_t towerIDBits, uint16_t timestampBits, uint16_t energyBits, uint16_t celltypeBits)
   {
-    auto dt = getDataRepresentation();
-    dt->mTowerID = tower;
-    dt->mTime = t;
-    dt->mEnergy = en;
-    dt->mCellStatus = status;
+    setEnergyEncoded(energyBits);
+    setTimestampEncoded(timestampBits);
+    setTowerIDEncoded(towerIDBits);
+    setChannelTypeEncoded(celltypeBits);
   }
 
-  auto getPackedTowerID() const { return getDataRepresentation()->mTowerID; }
-  auto getPackedTime() const { return getDataRepresentation()->mTime; }
-  auto getPackedEnergy() const { return getDataRepresentation()->mEnergy; }
-  auto getPackedCellStatus() const { return getDataRepresentation()->mCellStatus; }
+  /// \brief Get encoded bit representation of tower ID (for CTF)
+  /// \return Encoded bit representation
+  ///
+  /// Same as getTower - no compression applied for tower ID
+  uint16_t getTowerIDEncoded() const;
+
+  /// \brief Get encoded bit representation of timestamp (for CTF)
+  /// \return Encoded bit representation
+  ///
+  /// The time stamp is expressed in ns and has
+  /// a resolution of 1 ns. The time range which can
+  /// be stored is from -1023 to 1023 ns. In case the
+  /// range is exceeded the time is set to the limit
+  /// of the range.
+  uint16_t getTimeStampEncoded() const;
+
+  /// \brief Get encoded bit representation of energy (for CTF)
+  /// \return Encoded bit representation
+  ///
+  /// The energy range covered by the cell
+  /// is 0 - 250 GeV, with a resolution of
+  /// 0.0153 GeV. In case an energy exceeding
+  /// the limits is provided the energy is
+  /// set to the limits (0 in case of negative
+  /// energy, 250. in case of energies > 250 GeV)
+  uint16_t getEnergyEncoded() const;
+
+  /// \brief Get encoded bit representation of cell type (for CTF)
+  /// \return Encoded bit representation
+  uint16_t getCellTypeEncoded() const;
 
  private:
-  struct __attribute__((packed)) CellData {
-    uint16_t mTowerID : 15;   ///< bits 0-14   Tower ID
-    uint16_t mTime : 11;      ///< bits 15-25: Time (signed, can become negative after calibration)
-    uint16_t mEnergy : 14;    ///< bits 26-39: Energy
-    uint16_t mCellStatus : 2; ///< bits 40-41: Cell status
-    uint16_t mZerod : 6;      ///< bits 42-47: Zerod
-  };
+  /// \brief Set cell energy from encoded bit representation (from CTF)
+  /// \param energyBits Bit representation of energy
+  void setEnergyEncoded(uint16_t energyBits);
 
-  CellData* getDataRepresentation() { return reinterpret_cast<CellData*>(mCellWords); }
-  const CellData* getDataRepresentation() const { return reinterpret_cast<const CellData*>(mCellWords); }
+  /// \brief Set cell time from encoded bit representation (from CTF)
+  /// \param timestampBits Bit representation of timestamp
+  void setTimestampEncoded(uint16_t timestampBits);
 
-  char mCellWords[6]; ///< data word
+  /// \brief Set tower ID from encoded bit representation (from CTF)
+  /// \param towerIDBits Bit representation of towerID
+  void setTowerIDEncoded(uint16_t towerIDBits);
+
+  /// \brief Set cell type from encoded bit representation (from CTF)
+  /// \param channelTypeBits Bit representation of cell type
+  void setChannelTypeEncoded(uint16_t channelTypeBits);
+
+  short mTowerID = SHRT_MAX;                             ///< Tower ID
+  float mEnergy = FLT_MIN;                               ///< Energy
+  float mTimestamp = FLT_MIN;                            ///< Timestamp
+  ChannelType_t mChannelType = ChannelType_t::HIGH_GAIN; ///< Cell type
 
   ClassDefNV(Cell, 2);
 };

--- a/DataFormats/Detectors/EMCAL/src/Cell.cxx
+++ b/DataFormats/Detectors/EMCAL/src/Cell.cxx
@@ -19,9 +19,28 @@ using namespace o2::emcal;
 
 const float TIME_SHIFT = 600.,
             TIME_RANGE = 1500.,
-            TIME_RESOLUTION = TIME_RANGE / 2047.,
-            ENERGY_TRUNCATION = 250.,
-            ENERGY_RESOLUTION = ENERGY_TRUNCATION / 16383.;
+            TIME_RESOLUTION = TIME_RANGE / 2047.;
+namespace EnergyEncoding
+{
+namespace v0
+{
+const float
+  ENERGY_TRUNCATION = 250.,
+  ENERGY_RESOLUTION = ENERGY_TRUNCATION / 16383.;
+}
+
+namespace v1
+{
+const float
+  ENERGY_BITS = static_cast<float>(0x3FFF),
+  HGLGTRANSITION = o2::emcal::constants::EMCAL_HGLGTRANSITION * o2::emcal::constants::EMCAL_ADCENERGY,
+  ENERGY_TRUNCATION = 250.,
+  ENERGY_RESOLUTION_LG = (ENERGY_TRUNCATION - HGLGTRANSITION) / ENERGY_BITS,
+  ENERGY_RESOLUTION_HG = HGLGTRANSITION / ENERGY_BITS,
+  ENERGY_RESOLUTION_TRU = ENERGY_TRUNCATION / ENERGY_BITS,
+  ENERGY_RESOLUTION_LEDMON = ENERGY_TRUNCATION / ENERGY_BITS;
+}
+} // namespace EnergyEncoding
 
 Cell::Cell(short tower, float energy, float timestamp, ChannelType_t ctype) : mTowerID(tower), mEnergy(energy), mTimestamp(timestamp), mChannelType(ctype)
 {
@@ -48,13 +67,7 @@ uint16_t Cell::getTimeStampEncoded() const
 
 uint16_t Cell::getEnergyEncoded() const
 {
-  double truncatedEnergy = mEnergy;
-  if (truncatedEnergy < 0.) {
-    truncatedEnergy = 0.;
-  } else if (truncatedEnergy > ENERGY_TRUNCATION) {
-    truncatedEnergy = ENERGY_TRUNCATION;
-  }
-  return static_cast<int16_t>(std::round(truncatedEnergy / ENERGY_RESOLUTION));
+  return encodeEnergyV0(mEnergy);
 }
 
 uint16_t Cell::getCellTypeEncoded() const
@@ -64,7 +77,7 @@ uint16_t Cell::getCellTypeEncoded() const
 
 void Cell::setEnergyEncoded(uint16_t energyBits)
 {
-  mEnergy = energyBits * ENERGY_RESOLUTION;
+  mEnergy = decodeEnergyV0(energyBits);
 }
 
 void Cell::setTimestampEncoded(uint16_t timestampBits)
@@ -86,6 +99,85 @@ void Cell::truncate()
 {
   setEnergyEncoded(getEnergyEncoded());
   setTimestampEncoded(getTimeStampEncoded());
+}
+
+uint16_t Cell::encodeEnergyV0(float energy)
+{
+  auto truncatedEnergy = energy;
+  if (truncatedEnergy < 0.) {
+    truncatedEnergy = 0.;
+  } else if (truncatedEnergy > EnergyEncoding::v0::ENERGY_TRUNCATION) {
+    truncatedEnergy = EnergyEncoding::v0::ENERGY_TRUNCATION;
+  }
+  return static_cast<int16_t>(std::round(truncatedEnergy / EnergyEncoding::v0::ENERGY_RESOLUTION));
+}
+
+uint16_t Cell::encodeEnergyV1(float energy, ChannelType_t celltype)
+{
+  double truncatedEnergy = energy;
+  if (truncatedEnergy < 0.) {
+    truncatedEnergy = 0.;
+  } else if (truncatedEnergy > EnergyEncoding::v1::ENERGY_TRUNCATION) {
+    truncatedEnergy = EnergyEncoding::v1::ENERGY_TRUNCATION;
+  }
+  float resolutionApplied = 0., energyOffset = 0.;
+  switch (celltype) {
+    case ChannelType_t::HIGH_GAIN: {
+      resolutionApplied = EnergyEncoding::v1::ENERGY_RESOLUTION_HG;
+      break;
+    }
+    case ChannelType_t::LOW_GAIN: {
+      resolutionApplied = EnergyEncoding::v1::ENERGY_RESOLUTION_LG;
+      energyOffset = EnergyEncoding::v1::HGLGTRANSITION;
+      break;
+    }
+    case ChannelType_t::TRU: {
+      resolutionApplied = EnergyEncoding::v1::ENERGY_RESOLUTION_TRU;
+      break;
+    }
+    case ChannelType_t::LEDMON: {
+      resolutionApplied = EnergyEncoding::v1::ENERGY_RESOLUTION_LEDMON;
+      break;
+    }
+  }
+  return static_cast<uint16_t>(std::round((truncatedEnergy - energyOffset) / resolutionApplied));
+};
+
+uint16_t Cell::V0toV1(uint16_t energyBits, ChannelType_t celltype)
+{
+  auto decodedEnergy = decodeEnergyV0(energyBits);
+  return encodeEnergyV1(decodedEnergy, celltype);
+}
+
+float Cell::decodeEnergyV0(uint16_t energyBits)
+{
+  return static_cast<float>(energyBits) * EnergyEncoding::v0::ENERGY_RESOLUTION;
+}
+
+float Cell::decodeEnergyV1(uint16_t energyBits, ChannelType_t celltype)
+{
+  float resolutionApplied = 0.,
+        energyOffset = 0.;
+  switch (celltype) {
+    case ChannelType_t::HIGH_GAIN: {
+      resolutionApplied = EnergyEncoding::v1::ENERGY_RESOLUTION_HG;
+      break;
+    }
+    case ChannelType_t::LOW_GAIN: {
+      resolutionApplied = EnergyEncoding::v1::ENERGY_RESOLUTION_LG;
+      energyOffset = EnergyEncoding::v1::HGLGTRANSITION;
+      break;
+    }
+    case ChannelType_t::TRU: {
+      resolutionApplied = EnergyEncoding::v1::ENERGY_RESOLUTION_TRU;
+      break;
+    }
+    case ChannelType_t::LEDMON: {
+      resolutionApplied = EnergyEncoding::v1::ENERGY_RESOLUTION_LEDMON;
+      break;
+    }
+  }
+  return (static_cast<float>(energyBits) * resolutionApplied) + energyOffset;
 }
 
 void Cell::PrintStream(std::ostream& stream) const

--- a/DataFormats/Detectors/EMCAL/src/Cell.cxx
+++ b/DataFormats/Detectors/EMCAL/src/Cell.cxx
@@ -23,23 +23,19 @@ const float TIME_SHIFT = 600.,
             ENERGY_TRUNCATION = 250.,
             ENERGY_RESOLUTION = ENERGY_TRUNCATION / 16383.;
 
-Cell::Cell()
+Cell::Cell(short tower, float energy, float timestamp, ChannelType_t ctype) : mTowerID(tower), mEnergy(energy), mTimestamp(timestamp), mChannelType(ctype)
 {
-  memset(mCellWords, 0, 6);
 }
 
-Cell::Cell(short tower, float energy, float time, ChannelType_t ctype)
+uint16_t Cell::getTowerIDEncoded() const
 {
-  memset(mCellWords, 0, 6);
-  setTower(tower);
-  setTimeStamp(time);
-  setEnergy(energy);
-  setType(ctype);
+  return mTowerID;
 }
 
-void Cell::setTimeStamp(float timestamp)
+uint16_t Cell::getTimeStampEncoded() const
 {
   // truncate:
+  auto timestamp = mTimestamp;
   const float TIME_MIN = -1. * TIME_SHIFT,
               TIME_MAX = TIME_RANGE - TIME_SHIFT;
   if (timestamp < TIME_MIN) {
@@ -47,28 +43,49 @@ void Cell::setTimeStamp(float timestamp)
   } else if (timestamp > TIME_MAX) {
     timestamp = TIME_MAX;
   }
-  getDataRepresentation()->mTime = static_cast<uint16_t>(std::round((timestamp + TIME_SHIFT) / TIME_RESOLUTION));
+  return static_cast<uint16_t>(std::round((timestamp + TIME_SHIFT) / TIME_RESOLUTION));
 }
 
-float Cell::getTimeStamp() const
+uint16_t Cell::getEnergyEncoded() const
 {
-  return (static_cast<float>(getDataRepresentation()->mTime) * TIME_RESOLUTION) - TIME_SHIFT;
-}
-
-void Cell::setEnergy(float energy)
-{
-  double truncatedEnergy = energy;
+  double truncatedEnergy = mEnergy;
   if (truncatedEnergy < 0.) {
     truncatedEnergy = 0.;
   } else if (truncatedEnergy > ENERGY_TRUNCATION) {
     truncatedEnergy = ENERGY_TRUNCATION;
   }
-  getDataRepresentation()->mEnergy = static_cast<int16_t>(std::round(truncatedEnergy / ENERGY_RESOLUTION));
+  return static_cast<int16_t>(std::round(truncatedEnergy / ENERGY_RESOLUTION));
 }
 
-float Cell::getEnergy() const
+uint16_t Cell::getCellTypeEncoded() const
 {
-  return static_cast<float>(getDataRepresentation()->mEnergy) * ENERGY_RESOLUTION;
+  return static_cast<uint16_t>(mChannelType);
+}
+
+void Cell::setEnergyEncoded(uint16_t energyBits)
+{
+  mEnergy = energyBits * ENERGY_RESOLUTION;
+}
+
+void Cell::setTimestampEncoded(uint16_t timestampBits)
+{
+  mTimestamp = (timestampBits * TIME_RESOLUTION) - TIME_SHIFT;
+}
+
+void Cell::setTowerIDEncoded(uint16_t towerIDBits)
+{
+  mTowerID = towerIDBits;
+}
+
+void Cell::setChannelTypeEncoded(uint16_t channelTypeBits)
+{
+  mChannelType = static_cast<ChannelType_t>(channelTypeBits);
+}
+
+void Cell::truncate()
+{
+  setEnergyEncoded(getEnergyEncoded());
+  setTimestampEncoded(getTimeStampEncoded());
 }
 
 void Cell::PrintStream(std::ostream& stream) const

--- a/Detectors/CTF/test/test_ctf_io_emcal.cxx
+++ b/Detectors/CTF/test/test_ctf_io_emcal.cxx
@@ -14,6 +14,7 @@
 #define BOOST_TEST_DYN_LINK
 #include <boost/test/unit_test.hpp>
 #include "CommonUtils/NameConf.h"
+#include "DataFormatsEMCAL/Constants.h"
 #include "EMCALReconstruction/CTFCoder.h"
 #include "DataFormatsEMCAL/CTF.h"
 #include "Framework/Logger.h"
@@ -41,8 +42,32 @@ BOOST_AUTO_TEST_CASE(CTFTest)
     while (tower < 17665) {
       float timeCell = gRandom->Rndm() * 1500 - 600.;
       float en = gRandom->Rndm() * 250.;
-      int stat = gRandom->Integer(5);
-      cells.emplace_back(tower, en, timeCell, (ChannelType_t)stat);
+      // In case of cell type 3 cases must be distinguished (FEE, LEDMON, TRU)
+      // In case the cell is a FEE cell the cell type is correlated with the energy
+      int readoutsource = gRandom->Integer(3); //
+      ChannelType_t chantype = ChannelType_t::HIGH_GAIN;
+      switch (readoutsource) {
+        case 0: {
+          // Cell is a FEE cell, determine cell type according to HGLG transition
+          const auto ENHGLG = o2::emcal::constants::EMCAL_HGLGTRANSITION * o2::emcal::constants::EMCAL_ADCENERGY;
+          if (en >= ENHGLG) {
+            chantype = ChannelType_t::LOW_GAIN;
+          } else {
+            chantype = ChannelType_t::HIGH_GAIN;
+          }
+          break;
+        }
+        case 1:
+          chantype = ChannelType_t::LEDMON;
+          break;
+        case 2:
+          chantype = ChannelType_t::TRU;
+          break;
+
+        default:
+          break;
+      }
+      cells.emplace_back(tower, en, timeCell, chantype);
       tower += 1 + gRandom->Integer(100);
     }
     uint32_t trigBits = gRandom->Integer(0xFFFFFFFF); // will be converted internally to uint16_t by the coder
@@ -126,9 +151,9 @@ BOOST_AUTO_TEST_CASE(CTFTest)
   for (size_t i = 0; i < cells.size(); i++) {
     const auto& cor = cells[i];
     const auto& cdc = cellsD[i];
-    BOOST_CHECK_EQUAL(cor.getPackedTowerID(), cdc.getPackedTowerID());
-    BOOST_CHECK_EQUAL(cor.getPackedTime(), cdc.getPackedTime());
-    BOOST_CHECK_EQUAL(cor.getPackedEnergy(), cdc.getPackedEnergy());
-    BOOST_CHECK_EQUAL(cor.getPackedCellStatus(), cdc.getPackedCellStatus());
+    BOOST_CHECK_EQUAL(cor.getTowerIDEncoded(), cdc.getTowerIDEncoded());
+    BOOST_CHECK_EQUAL(cor.getTimeStampEncoded(), cdc.getTimeStampEncoded());
+    BOOST_CHECK_EQUAL(cor.getEnergyEncoded(), cdc.getEnergyEncoded());
+    BOOST_CHECK_EQUAL(cor.getCellTypeEncoded(), cdc.getCellTypeEncoded());
   }
 }

--- a/Detectors/CTF/test/test_ctf_io_emcal.cxx
+++ b/Detectors/CTF/test/test_ctf_io_emcal.cxx
@@ -65,6 +65,7 @@ BOOST_AUTO_TEST_CASE(CTFTest)
           break;
 
         default:
+          std::cerr << "Unknown type" << std::endl;
           break;
       }
       cells.emplace_back(tower, en, timeCell, chantype);

--- a/Detectors/EMCAL/reconstruction/include/EMCALReconstruction/CTFCoder.h
+++ b/Detectors/EMCAL/reconstruction/include/EMCALReconstruction/CTFCoder.h
@@ -188,7 +188,7 @@ o2::ctf::CTFIOSize CTFCoder::decode(const CTF::base& ec, VTRG& trigVec, VCELL& c
     firstEntry = cellVec.size();
 
     for (uint16_t ic = 0; ic < entries[itrig]; ic++) {
-      cell.setPacked(tower[cellCount], cellTime[cellCount], energy[cellCount], status[cellCount]);
+      cell.initialiseFromEncoded(tower[cellCount], cellTime[cellCount], energy[cellCount], status[cellCount]);
       cellVec.emplace_back(cell);
       cellCount++;
     }

--- a/Detectors/EMCAL/reconstruction/include/EMCALReconstruction/CTFCoder.h
+++ b/Detectors/EMCAL/reconstruction/include/EMCALReconstruction/CTFCoder.h
@@ -166,6 +166,13 @@ o2::ctf::CTFIOSize CTFCoder::decode(const CTF::base& ec, VTRG& trigVec, VCELL& c
   trigVec.reserve(header.nTriggers);
   status.reserve(header.nCells);
 
+  /*
+  Bool_t isVersion1 = kFALSE;
+  if (header.majorVersion == 0 && header.minorVersion == 1) {
+    isVersion1 = kTRUE;
+  }
+  */
+
   uint32_t firstEntry = 0, cellCount = 0;
   o2::InteractionRecord ir(header.firstBC, header.firstOrbit);
   bool checkIROK = (mBCShift == 0); // need to check if CTP offset correction does not make the local time negative ?
@@ -188,6 +195,13 @@ o2::ctf::CTFIOSize CTFCoder::decode(const CTF::base& ec, VTRG& trigVec, VCELL& c
     firstEntry = cellVec.size();
 
     for (uint16_t ic = 0; ic < entries[itrig]; ic++) {
+      /*
+      uint16_t packedEnergy = energy[cellCount];
+      if (isVersion1) {
+        packedEnergy = o2::emcal::Cell::V0toV1(energy[cellCount], status[cellCount]);
+      }
+      cell.initialiseFromEncoded(tower[cellCount], cellTime[cellCount], packedEnergy, status[cellCount]);
+      */
       cell.initialiseFromEncoded(tower[cellCount], cellTime[cellCount], energy[cellCount], status[cellCount]);
       cellVec.emplace_back(cell);
       cellCount++;

--- a/Detectors/EMCAL/reconstruction/include/EMCALReconstruction/CTFHelper.h
+++ b/Detectors/EMCAL/reconstruction/include/EMCALReconstruction/CTFHelper.h
@@ -201,8 +201,8 @@ class CTFHelper
   {
    public:
     using _Iter<Iter_towerID, Cell, uint16_t>::_Iter;
-    value_type operator*() const { return mData[mIndex].getPackedTowerID(); }
-    value_type operator[](difference_type i) const { return mData[mIndex + i].getPackedTowerID(); }
+    value_type operator*() const { return mData[mIndex].getTowerIDEncoded(); }
+    value_type operator[](difference_type i) const { return mData[mIndex + i].getTowerIDEncoded(); }
   };
 
   //_______________________________________________
@@ -210,8 +210,8 @@ class CTFHelper
   {
    public:
     using _Iter<Iter_time, Cell, uint16_t>::_Iter;
-    value_type operator*() const { return mData[mIndex].getPackedTime(); }
-    value_type operator[](difference_type i) const { return mData[mIndex + i].getPackedTime(); }
+    value_type operator*() const { return mData[mIndex].getTimeStampEncoded(); }
+    value_type operator[](difference_type i) const { return mData[mIndex + i].getTimeStampEncoded(); }
   };
 
   //_______________________________________________
@@ -219,8 +219,8 @@ class CTFHelper
   {
    public:
     using _Iter<Iter_energy, Cell, uint16_t>::_Iter;
-    value_type operator*() const { return mData[mIndex].getPackedEnergy(); }
-    value_type operator[](difference_type i) const { return mData[mIndex + i].getPackedEnergy(); }
+    value_type operator*() const { return mData[mIndex].getEnergyEncoded(); }
+    value_type operator[](difference_type i) const { return mData[mIndex + i].getEnergyEncoded(); }
   };
 
   //_______________________________________________
@@ -228,8 +228,8 @@ class CTFHelper
   {
    public:
     using _Iter<Iter_status, Cell, uint8_t>::_Iter;
-    value_type operator*() const { return mData[mIndex].getPackedCellStatus(); }
-    value_type operator[](difference_type i) const { return mData[mIndex + i].getPackedCellStatus(); }
+    value_type operator*() const { return mData[mIndex].getCellTypeEncoded(); }
+    value_type operator[](difference_type i) const { return mData[mIndex + i].getCellTypeEncoded(); }
   };
 
   //<<< =========================== ITERATORS ========================================

